File: src/main/java/me/fourteendoggo/xcore/XCore.java
package me.fourteendoggo.xcore;

import co.aikar.commands.BukkitCommandManager;
import co.aikar.commands.InvalidCommandArgument;
import me.fourteendoggo.xcore.commands.HomeCommand;
import me.fourteendoggo.xcore.commands.ReloadCommand;
import me.fourteendoggo.xcore.commands.vanish.VanishCommand;
import me.fourteendoggo.xcore.commands.vanish.VanishManager;
import me.fourteendoggo.xcore.inventory.InventoryManager;
import me.fourteendoggo.xcore.listeners.PlayerListener;
import me.fourteendoggo.xcore.storage.Storage;
import me.fourteendoggo.xcore.storage.UserManager;
import me.fourteendoggo.xcore.user.Home;
import me.fourteendoggo.xcore.user.User;
import me.fourteendoggo.xcore.utils.Lang;
import me.fourteendoggo.xcore.utils.Reloadable;
import me.fourteendoggo.xcore.utils.Settings;
import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.PluginManager;
import org.bukkit.plugin.java.JavaPlugin;

import java.util.Arrays;
import java.util.List;
import java.util.function.Consumer;

public class XCore extends JavaPlugin {
    private boolean reloading;
    private Storage storage;
    private UserManager userManager;
    private VanishManager vanishManager;
    private InventoryManager inventoryManager;
    private List<Reloadable> reloadableComponents;

    @Override
    public void onLoad() {
        reloading = !Bukkit.getWorlds().isEmpty();
    }

    @Override
    @SuppressWarnings("deprecation")
    public void onEnable() {
        saveDefaultConfig();

        Settings.loadFromDisk(this);
        Lang.loadFromDisk(this);

        storage = new Storage(getLogger());
        userManager = new UserManager(storage);
        vanishManager = new VanishManager(this);
        inventoryManager = new InventoryManager();

        PluginManager pm = Bukkit.getPluginManager();
        pm.registerEvents(new PlayerListener(this, vanishManager), this);
        pm.registerEvents(inventoryManager, this);

        BukkitCommandManager commandManager = new BukkitCommandManager(this);
        commandManager.enableUnstableAPI("help");

        commandManager.getCommandContexts().registerIssuerOnlyContext(User.class, context -> {
            if (context.getSender() instanceof Player player) {
                return userManager.getUser(player.getUniqueId());
            }
            throw new InvalidCommandArgument("Console cannot execute this command");
        });

        commandManager.getCommandContexts().registerContext(Home.class, context -> {
            String name = context.popFirstArg();
            Player player = context.getPlayer();
            User user = userManager.getUser(player.getUniqueId());
            return user.getData().removeHome(name);
        });

        commandManager.registerCommand(new HomeCommand(this));
        commandManager.registerCommand(new ReloadCommand(this));
        commandManager.registerCommand(new VanishCommand(vanishManager));

        reloadableComponents = Arrays.asList(
                Lang::reload, Settings::reload, vanishManager
        );

        if (reloading && !Bukkit.getOnlinePlayers().isEmpty()) {
            getLogger().info("Reload detected, reloading all players...");
            // renew states in case of the plugins jar being replaced
            for (User user : userManager) {
                Player player = user.getPlayer();
                userManager.loadIfAbsent(player.getUniqueId());
                vanishManager.vanishIfFlagged(user);
            }
        }

        getLogger().info(reloading ? "XCore has been reloaded" : "XCore has been enabled");
    }

    @Override
    public void onDisable() {
        ifNotNull(vanishManager, VanishManager::destroyState);
        ifNotNull(userManager, UserManager::saveAllBlocking);
        ifNotNull(storage, Storage::disconnect);

        getLogger().info("XCore has been disabled");
    }

    /* avoid null pointer exceptions in onDisable due to onEnable not fully being called */
    private <T> void ifNotNull(T instance, Consumer<T> function) {
        if (instance != null) {
            function.accept(instance);
        }
    }

    public void reload() {
        reloadConfig();
        reloadableComponents.forEach(Reloadable::reload);
        getLogger().info("XCore has been reloaded");
    }

    public Storage getStorage() {
        return storage;
    }

    public UserManager getUserManager() {
        return userManager;
    }

    public InventoryManager getInventoryManager() {
        return inventoryManager;
    }
}

File: src/main/java/me/fourteendoggo/xcore/XCore.kt
package me.fourteendoggo.xcore

import co.aikar.commands.BukkitCommandManager
import co.aikar.commands.InvalidCommandArgument
import me.fourteendoggo.xcore.commands.HomeCommand
import me.fourteendoggo.xcore.commands.ReloadCommand
import me.fourteendoggo.xcore.commands.vanish.VanishCommand
import me.fourteendoggo.xcore.commands.vanish.VanishManager
import me.fourteendoggo.xcore.inventory.InventoryManager
import me.fourteendoggo.xcore.listeners.PlayerListener
import me.fourteendoggo.xcore.listeners.SkillsListener
import me.fourteendoggo.xcore.skills.FarmingSkill
import me.fourteendoggo.xcore.skills.MiningSkill
import me.fourteendoggo.xcore.skills.SkillsManager
import me.fourteendoggo.xcore.skills.WoodCuttingSkill
import me.fourteendoggo.xcore.storage.Storage
import me.fourteendoggo.xcore.storage.UserManager
import me.fourteendoggo.xcore.user.Home
import me.fourteendoggo.xcore.user.User
import me.fourteendoggo.xcore.utils.LangKey
import me.fourteendoggo.xcore.utils.Reloadable
import me.fourteendoggo.xcore.utils.Settings
import org.bukkit.Bukkit
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin

class XCore : JavaPlugin() {
    private var reloading = false
    lateinit var storage: Storage private set
    lateinit var userManager: UserManager private set
    private lateinit var vanishManager: VanishManager
    lateinit var inventoryManager: InventoryManager private set
    private lateinit var reloadableComponents: Array<Reloadable>

    override fun onLoad() {
        reloading = Bukkit.getWorlds().isNotEmpty()
    }

    override fun onEnable() {
        saveDefaultConfig()

        Settings.loadFromDisk(this)
        LangKey.loadFromDisk(this)

        storage = Storage(logger)
        userManager = UserManager(storage)
        vanishManager = VanishManager(this)
        inventoryManager = InventoryManager()

        Bukkit.getPluginManager().apply {
            registerEvents(PlayerListener(this@XCore, vanishManager), this@XCore)
            registerEvents(inventoryManager, this@XCore)

            val skillsManager = SkillsManager().apply {
                addSkills(FarmingSkill, WoodCuttingSkill, MiningSkill)
            }
            registerEvents(SkillsListener(userManager, skillsManager), this@XCore)
        }

        BukkitCommandManager(this).apply {
            @Suppress("DEPRECATION")
            enableUnstableAPI("help")

            commandContexts.registerIssuerOnlyContext(User::class.java) { ctx ->
                val player = (ctx.sender as? Player) ?: throw InvalidCommandArgument("Console cannot execute this command")
                return@registerIssuerOnlyContext userManager[player.uniqueId]
            }

            commandContexts.registerContext(Home::class.java) { ctx ->
                val name = ctx.popFirstArg()
                val user = userManager[ctx.player.uniqueId]
                return@registerContext user.data.removeHome(name)
            }

            registerCommand(HomeCommand(this@XCore))
            registerCommand(ReloadCommand(this@XCore))
            registerCommand(VanishCommand(vanishManager))
        }

        reloadableComponents = arrayOf(
            LangKey.Companion, Settings, vanishManager,
        )

        if (reloading && Bukkit.getOnlinePlayers().isNotEmpty()) {
            logger.info("Reload detected, reloading all players...")
            // renew states in case of the plugins jar being replaced
            for (user in userManager) {
                // FIXME: why is this necessary?
                userManager.unloadUser(user.id)
                vanishManager.vanishIfFlagged(user)
            }
        }

        val status = if (reloading) "reloaded" else "enabled"
        logger.info("$name has been $status")
    }

    override fun onDisable() {
        vanishManager.destroyState()
        userManager.saveAllBlocking()
        storage.disconnect()

        logger.info("$name has been disabled")
    }

    fun reload() {
        reloadConfig()
        reloadableComponents.forEach(Reloadable::reload)
        logger.info("$name has been reloaded")
    }
}